pkg/util/containers/v2/metrics/collector_containerd.go

```diff
@@ -0,0 +1,401 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// +build containerd
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	wstats "github.com/Microsoft/hcsshim/cmd/containerd-shim-runhcs-v1/stats"
+	v1 "github.com/containerd/cgroups/stats/v1"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/typeurl"
+	"github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/providers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/system"
+)
+
+const (
+	containerdCollectorID    = "containerd"
+	containerdCacheKeyPrefix = "containerd-stats-"
+	containerdCacheTTL       = 10 * time.Second
+)
+
+func init() {
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       containerdCollectorID,
+		priority: 1, // Less than the "system" collector, so we can rely on cgroups directly if possible
+		runtimes: []string{RuntimeNameContainerd},
+		factory: func() (Collector, error) {
+			return newContainerdCollector()
+		},
+	})
+}
+
+type containerdCollector struct {
+	client         cutil.ContainerdItf
+	lastScrapeTime time.Time
+}
+
+func newContainerdCollector() (*containerdCollector, error) {
+	if !config.IsFeaturePresent(config.Containerd) {
+		return nil, ErrPermaFail
+	}
+
+	client, err := cutil.GetContainerdUtil()
```

Comment:
Might be risky to `PermaFail` directly. I know our retrier is not great in terms of status but it may be useful to retry a few times at least or to `PermaFail` when Retrier is also `PermaFail`

Reply:
My understanding is that `cutil.GetContainerdUtil` already handles retries. The retrying logic can only return 2 untyped errors: `retry number exceeded` and `retryer not initialised`. In the case of `cutil.GetContainerdUtil` the second is not possible because it takes care of initializing the retrier. Did I miss something?

Reply:
@vboulineau any reason why the collector machinery doesn't rely on the same error types in the retrier package? it would make this way easier to handle

Reply:
The collectors would just return the error returned by the retriers and `retryCollectors` will do the rest.

Reply:
That's because the retrier is not used everywhere (see `github.com/cenkalti/backoff` - and is not awesome code TBH), so I don't think it's clean to force dependencies toward the `retrier` error types.

However we can add a small wrapper to ease the handling of retrier errors in the `metrics` package.

@davidor If `ContainerdUtil.connect()` returns an error the retrier will return a wrapped error with `r.status = FailWillRetry`

Reply:
@vboulineau I misunderstood the retrier logic. I fixed the code, I think it should be OK now.

Reply:
@davidor if you could write a small util that converts the retry pkg errors into metrics errors we can use it in the ecs fargate collector as well (and probably other collectors)

Reply:
@ahmed-mez @vboulineau I added a new commit with a helper to convert errors.

Reply:
thanks!